### PR TITLE
SA-1620: Group by table comparison filters

### PR DESCRIPTION
**Pull request by PatrickSison**
### What Changed

- Group By table swaps for CompareByGroupByTable (if there are comparison filters)
- Column added to comparison filter
- Added new useSparkPostQueries

### How To Test

- ON report builder with compare by enabled, add at least 2 comparison filters
 - Go to group by table and verify table works as expected
 - Verify that clicking filter links still adds filters to table
 - Verify that table is still sortable by column
 - Verify that default sort is by group by column (no longer first column)
- Block one of the comparison filters request from going through (and verify error state)
 - Block all of the comparison filters requests from going through (and verify error state)

### To Do

- [ ] Add testing
 - [ ] Disable option for group by if select column is used (or maybe just disable 2nd row if it's the same?)

### CODE REVIEW

File: src/helpers/api/metrics.js [hunk 1]
```diff
@@ -33,10 +33,10 @@ export function getTemplatesMetrics(params) {
   };
 }
 
-export function getDeliverabilityMetrics(params) {
+export function getDeliverabilityMetrics(params, path) {
   return {
     method: 'GET',
-    url: DELIVERABILITY_BASE_URL,
+    url: `${DELIVERABILITY_BASE_URL}${path ? `/${path}` : ''}`,
```
Comment (nicklemmon) [inline src/helpers/api/metrics.js:39]: Doesn't the `params` argument handle this rather than interpolating here?

File: src/pages/reportBuilder/ReportBuilder.js [hunk 2]
```diff
@@ -21,8 +21,10 @@ import {
   AggregatedMetrics,
   CompareByAggregatedMetrics,
   ReportOptions,
-  ReportTable,
+  GroupByTable,
+  // CompareByGroupByTable,
```
Comment (nicklemmon) [inline src/pages/reportBuilder/ReportBuilder.js:25]: Can this be removed?

File: src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  const {
+    selectors: { selectSummaryMetricsProcessed: metrics },
+  } = useReportBuilderContext();
+  const hasSubaccounts = useSelector(hasSubaccountsSelector);
+  const subaccounts = useSelector(state => state.subaccounts.list);
+  const group = GROUP_BY_CONFIG[groupBy];
+
+  const tableData = _.flatten(data);
+
+  const getColumnHeaders = () => {
```
Comment (nicklemmon) [inline src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js:38]: I'm not seeing the default sort within the table?
Screen Shot 2020-12-11 at 1 43 02 PM

File: src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+      return null;
+    }
+
+    if (status === 'loading') {
+      return <PanelLoading minHeight="250px" />;
+    }
+
+    if (!tableData.length) {
+      return (
+        <Panel.LEGACY>
```
Comment (nicklemmon) [inline src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js:108]: No need to use `Panel.LEGACY` within this feature.

File: src/pages/reportBuilder/components/GroupByTable/GroupByTable.js [hunk 5]
```diff
@@ -23,7 +23,7 @@ const tableWrapper = props => {
   );
 };
 
-export const ReportTable = () => {
+export const GroupByTable = () => {
```
Comment (nicklemmon) [inline src/pages/reportBuilder/components/GroupByTable/GroupByTable.js:26]: 👏  Much clearer

File: src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
+import React from 'react';
```
Comment (nicklemmon) [inline src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js:1]: We should have some Cypress testing around all of this? Probably useful to do testing around the multiple requests and their arguments as well as the user's interaction with the table when comparisons are enabled.

File: src/pages/reportBuilder/components/GroupByTable/useGroupByTable.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
+import { useState, useMemo } from 'react';
```
Comment (nicklemmon) [inline src/pages/reportBuilder/components/GroupByTable/useGroupByTable.js:1]: This could all be tested via the feature itself rather than Jest unit testing, though unit testing might be a little easier to test independently (as well as edge cases).

We have this installed:
https://github.com/testing-library/react-hooks-testing-library

Up to you how it should be handled.

**Review by nicklemmon — CHANGES_REQUESTED** [review]
@PatrickSison Seems to work well overall. I have a few comments I've left, plus I don't see any testing to support these changes so that's a concern.

File: src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js (hunk 3 above)
Comment (PatrickSison) [reply src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js:38]: I think default sort is the sort is uses if no sort is selected. Active sorting works otherwise

File: src/helpers/api/metrics.js (hunk 1 above)
Comment (PatrickSison) [reply src/helpers/api/metrics.js:39]: Was on the fence with how to handle this, but also it uses the paths `/deliverability/${groupByKey}` to actually route to the appropriate break down table. It feels separate from the rest of the params, but we could pass it in through params and deconstruct just this variable, and send the rest down to the API

Comment (nicklemmon) [reply src/helpers/api/metrics.js:39]: Yeah I think that would be better in my mind - make this as generic/simple as possible and keep complexity elsewhere.

File: src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+import { useReportBuilderContext } from '../../context/ReportBuilderContext';
+import AddFilterLink from '../AddFilterLink';
+import { ApiErrorBanner } from 'src/components';
+
+import styles from './ReportTable.module.scss';
+import useGroupByTable from './useGroupByTable';
+
+const tableWrapper = props => {
+  return (
+    <Panel.LEGACY>
```
Comment (nicklemmon) [inline src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js:22]: Same here, no need to use `Panel.LEGACY`

File: src/hooks/useSparkPostQuery/useSparkPostQuery.js [hunk 9]
```diff
@@ -5,6 +5,33 @@ import { refresh, logout } from 'src/actions/auth';
...
+
+export function useSparkPostQueries(queries, config = {}, queryKey) {
+  const queryCache = useQueryCache();
+  const auth = useSelector(state => state.auth);
+  const dispatch = useDispatch();
+
+  const queryRequests = () =>
+    queries.map(queryFn => {
+      const { url, method, params, headers } = queryFn();
+      return defaultQuery(url, { method, params, headers, auth });
```
Comment (nicklemmon) [inline src/hooks/useSparkPostQuery/useSparkPostQuery.js:18]: You shouldn't have to pass things in to the `defaultQuery` manually? From what I understand, this is how it works:
- A `queryKey` can contain lots of request-related information (method, URL, params) which is passed to each instance of `useQuery`
- `useQuery` takes that `queryKey` and passes it to our axios instance, `sparkPostRequest` - our custom `defaultQuery` destructures the `queryKey` and does that for us
- We pass the `defaultQuery` to the `QueryCache` component so that it is always there, consuming/handling passed-in query keys without us having to manually reference it

File: src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+      return <PanelLoading minHeight="250px" />;
+    }
+
+    if (!tableData.length) {
+      return (
+        <Panel>
+          <Empty message="There is no data to display" />
+        </Panel>
+      );
+    }
```
Comment (nicklemmon) [inline src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js:126]: It would be worth adding some testing for the error and empty state scenarios.

File: src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+    const subaccount = subaccounts.find(({ id }) => {
+      return id === subaccountId;
+    });
+
+    const value = subaccount
+      ? `${subaccount?.name} (ID ${subaccount?.id})`
+      : `Subaccount ${subaccountId}`;
+    return { type: 'Subaccount', value, id: subaccountId };
+  };
```
Comment (nicklemmon) [inline src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js:78]: Screen Shot 2020-12-14 at 1 12 30 PM

Appear to be some issues with rendering subaccounts.

Comment (nicklemmon) [reply src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js:78]: When clicking on these links that does appear to cause some invalid queries to be passed to the API. Fortunately error scenarios handle this well.

**Review by nicklemmon — CHANGES_REQUESTED** [review]
@PatrickSison taking another pass and I think I caught one bug. Other feedback is minor, though I do think some additional testing of edge cases is warranted.

File: src/hooks/useSparkPostQuery/useSparkPostQuery.js (hunk 9 above)
Comment (PatrickSison) [reply src/hooks/useSparkPostQuery/useSparkPostQuery.js:18]: I think this is a limit of `useQuery` solved by `useQueries` in `react-query`. For this instance, we don't necessarily know the number of requests that could come in because of the comparisons. Regardless of the number of requests, these have to be joined into one promise as if one fails, the rest fails because the data can't be loaded altogether.

Comment (nicklemmon) [reply src/hooks/useSparkPostQuery/useSparkPostQuery.js:18]: That makes sense to me.

And in the not too distant future, we'll migrate to React Query 3, and then a lot of this should go away - though not all of it.

File: cypress/tests/integration/signals-analytics/analytics-report/reportBreakdownTable.spec.js [hunk 12]
```diff
@@ -459,27 +459,94 @@ if (IS_HIBANA_ENABLED) {
...
 
-      cy.get('table').within(() => {
-        cy.findAllByText('Fake Subaccount 1 (ID 101)').should('have.length', 3);
-        cy.findAllByText('Fake Subaccount 3 (ID 103)').should('have.length', 3);
+        cy.wait('@getWatchedDomains');
+        cy.wait('@getWatchedDomains');
+        cy.findAllByText('There is no data to display').should('be.visible');
+      });
+
+      it('will render an empty state the requests fail', () => {
```
Comment (nicklemmon) [inline cypress/tests/integration/signals-analytics/analytics-report/reportBreakdownTable.spec.js:507]: This is an error state, no? Not an empty state?

File: cypress/tests/integration/signals-analytics/analytics-report/reportBreakdownTable.spec.js [hunk 13]
```diff
@@ -459,27 +459,94 @@ if (IS_HIBANA_ENABLED) {
...
+
+        // Hmmmmmm
+        cy.wait('@getWatchedDomains');
+        cy.wait('@getWatchedDomains');
+        cy.wait('@getWatchedDomains');
+        cy.wait('@getWatchedDomains');
+        cy.wait('@getWatchedDomains');
+        cy.wait('@getWatchedDomains');
+        cy.wait('@getWatchedDomains');
+        cy.wait('@getWatchedDomains');
```
Comment (nicklemmon) [inline cypress/tests/integration/signals-analytics/analytics-report/reportBreakdownTable.spec.js:531]: Seems like a bug!

File: cypress/tests/integration/signals-analytics/analytics-report/reportBreakdownTable.spec.js [hunk 14]
```diff
@@ -459,27 +459,94 @@ if (IS_HIBANA_ENABLED) {
...
+
+        cy.stubRequest({
+          url: '/api/v1/metrics/deliverability/watched-domain**/*',
+          fixture: 'metrics/deliverability/watched-domain/200.get.json',
+          requestAlias: 'getWatchedDomains',
+        });
+
+        cy.findByRole('button', { name: 'Try Again' }).click();
+        cy.wait('@getWatchedDomains');
+        cy.wait('@getWatchedDomains');
```
Comment (nicklemmon) [inline cypress/tests/integration/signals-analytics/analytics-report/reportBreakdownTable.spec.js:544]: Since they occur simultaneously, probably use this:
```
cy.wait(['@getWatchedDomains', '@getWatchedDomains']);
```

Just helps things be a little bit more self documenting.

File: src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+      group.label === 'Subaccount'
+        ? getSubaccountFilter(filterKey)
+        : { type: group.label, value: filterKey };
+
+    const primaryCol = <AddFilterLink newFilter={newFilter} />;
+    const comparisonCol = <Tag>{row[comparisonType]}</Tag>;
+    const metricCols = metrics.map(({ key, unit }) => (
+      <Box textAlign="right" key={key}>
+        <Unit value={row[key] || '- - -'} unit={unit} />
+      </Box>
```
Comment (nicklemmon) [inline src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js:92]: This component should be implemented here. If it's global and re-usable, we can also use it in lots of spots for consistency's sake:
```jsx
function EmptyCell() {
  return (
    <>
      {/* hidden from screen reader users */}
      <span aria-hidden="true">---</span>
      {/* hidden from everyone else */}
      <ScreenReaderOnly>No Data</ScreenReaderOnly>
    </>
  );
}
```

File: src/pages/reportBuilder/components/GroupByTable/useGroupByTable.js [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+    return [reportOptions];
+  }
+
+  return comparisons.map(compareFilter => {
+    const filterType = REPORT_BUILDER_FILTER_KEY_MAP[compareFilter.type];
+    const comparedFilters = [
+      ...reportOptions.filters,
+      { AND: { [filterType]: { eq: [compareFilter] } } },
+    ];
+    return { ...reportOptions, filters: comparedFilters };
```
Comment (nicklemmon) [inline src/pages/reportBuilder/components/GroupByTable/useGroupByTable.js:25]: This could be tough to fix, but looks like it's possible to add the same filter multiple times through this table, resulting in an API error. It's a weird edge case, so I'm not sure how that should be handled:
Screen Shot 2020-12-16 at 10 13 31 AM

File: src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+  };
+
+  const TableContent = () => {
+    if (!group || metrics.length === 0) {
+      return null;
+    }
+
+    if (status === 'error') {
+      return (
+        <Panel>
```
Comment (nicklemmon) [inline src/pages/reportBuilder/components/GroupByTable/CompareByGroupByTable.js:105]: We would want to put this inside a `Panel.Section` for consistency and to avoid having a double border when we upgrade Matchbox (white line is there to show the alignment inconsistency):

Screen Shot 2020-12-16 at 10 15 18 AM

File: jsconfig.json [hunk 18]
```diff
@@ -0,0 +1,7 @@
+{
+  "compilerOptions": {
+    "allowSyntheticDefaultImports": false,
+    "baseUrl": "./"
+  },
+  "exclude": ["node_modules", "dist"]
```
Comment (nicklemmon) [inline jsconfig.json:6]: Our build folder is `build` rather than `dist` looks like.

File: src/hooks/useSparkPostQuery/useSparkPostQuery.js [hunk 19]
```diff
@@ -5,8 +5,35 @@ import { refresh, logout } from 'src/actions/auth';
 import { useRefreshToken } from 'src/helpers/http';
 import { showAlert } from 'src/actions/globalAlert';
 import { fetch as fetchAccount } from 'src/actions/account';
+import { defaultQuery } from 'src/helpers/api';
 
-export default function useSparkPostQuery(queryFn, config = {}) {
+export function useSparkPostQueries(queries, config = {}, queryKey) {
```
Comment (nicklemmon) [inline src/hooks/useSparkPostQuery/useSparkPostQuery.js:10]: Would you mind adding some tests to `useSparkPostQuery.spec.js` and some docs to `useSparkPostQuery.stories.mdx`?

I would think just a tiny bit of both would go a long way.

File: src/pages/reportBuilder/components/GroupByTable/useGroupByTable.js [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+    // Array of objects (unique group-by key)
+    const objectKeys = _.uniqBy(_.flatten(rawData), groupKey);
+
+    return objectKeys.reduce((acc, object) => {
+      const key = object[groupKey];
+      const newRows = comparisons.map((compareFilter, index) => {
+        const currentDataCompare = keyedData[index][key] || {};
+        return {
+          ...currentDataCompare,
+          [groupKey]: key, //Readd key (in case it's empty)
```
Comment (nicklemmon) [inline src/pages/reportBuilder/components/GroupByTable/useGroupByTable.js:77]: Typo here

File: cypress/tests/integration/signals-analytics/analytics-report/reportBreakdownTable.spec.js (hunk 13 above)
Comment (nicklemmon) [reply cypress/tests/integration/signals-analytics/analytics-report/reportBreakdownTable.spec.js:531]: Nevermind! This is right. It looks like there are two requests that are invoked via `useSparkPostQueries` each time it attempts the request, so that should total 8. This works too:

```js
cy.wait(['@getWatchedDomains', '@getWatchedDomains']);
cy.wait(['@getWatchedDomains', '@getWatchedDomains']);
cy.wait(['@getWatchedDomains', '@getWatchedDomains']);
cy.wait(['@getWatchedDomains', '@getWatchedDomains']);
```

**Review by nicklemmon — CHANGES_REQUESTED** [review]
@PatrickSison Left a few comments, though not a whole lot - I think this is almost ready.

File: cypress/tests/integration/signals-analytics/analytics-report/reportBreakdownTable.spec.js (hunk 13 above)
Comment (PatrickSison) [reply cypress/tests/integration/signals-analytics/analytics-report/reportBreakdownTable.spec.js:531]: That's a lot nicer

**Review by nicklemmon — APPROVED** [review]
@PatrickSison 👍 Looks good to me.
